ChopChop/Models/Recipe/Recipe/Recipe.swift

```diff
@@ -0,0 +1,204 @@
+import Foundation
+import GRDB
+
+/// Note there is no relationship between steps and ingredients after parsing stage
+class Recipe: FetchableRecord {
+    var id: Int64?
+    private(set) var name: String
+    private(set) var servings: Double
+    var recipeCategoryId: Int64?
+    private(set) var difficulty: Difficulty?
+    private(set) var steps: [RecipeStep]
+    private(set) var ingredients: [RecipeIngredient]
+
+    init(name: String, servings: Double = 1, difficulty: Difficulty? = nil, steps: [RecipeStep] = [], ingredients: [RecipeIngredient] = []) throws {
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+
+        guard servings > 0 else {
+            throw RecipeError.invalidServings
+        }
+        self.servings = servings
+        self.difficulty = difficulty
+        self.steps = steps
+        self.ingredients = ingredients
+        assert(checkRepresentation())
+    }
+
+    func updateName(_ name: String) throws {
+        assert(checkRepresentation())
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+        assert(checkRepresentation())
+    }
+
+    func updateRecipeCategory(_ recipeCategory: Int64) {
+        assert(checkRepresentation())
+        self.recipeCategoryId = recipeCategory
+        assert(checkRepresentation())
+    }
+
+    func updateDifficulty(_ difficulty: Difficulty) {
+        assert(checkRepresentation())
+        self.difficulty = difficulty
+        assert(checkRepresentation())
+    }
+
+    // step related functions
+    func addStep(_ addedStep: RecipeStep) {
+        assert(checkRepresentation())
+        steps.append(addedStep)
+        assert(checkRepresentation())
+    }
+
+    func removeStep(_ removedStep: RecipeStep) throws {
```

Comment:
Is the error useful? If not, maybe this method can just do nothing if the step is not existent